src/box/sql/mem.c

```diff
@@ -1001,6 +1099,173 @@ double_to_str0(struct Mem *mem)
 	return 0;
 }
 
+static inline int
+dec_to_int(struct Mem *mem)
+{
+	assert(mem->type == MEM_TYPE_DEC);
+	if (decimal_is_neg(&mem->u.d)) {
+		int64_t i;
+		if (decimal_to_int64(&mem->u.d, &i) == NULL)
+			return -1;
+		assert(i < 0);
+		mem->u.i = i;
+		mem->type = MEM_TYPE_INT;
+		mem->flags = 0;
+		return 0;
+	}
+	uint64_t u;
+	if (decimal_to_uint64(&mem->u.d, &u) == NULL)
+		return -1;
+	mem->u.u = u;
+	mem->type = MEM_TYPE_UINT;
+	mem->flags = 0;
+	return 0;
+}
+
+static inline int
+dec_to_int_precise(struct Mem *mem)
+{
+	assert(mem->type == MEM_TYPE_DEC);
+	if (!decimal_is_int(&mem->u.d))
+		return -1;
+	return dec_to_int(mem);
+}
+
+static inline int
+dec_to_int_forced(struct Mem *mem)
+{
+	assert(mem->type == MEM_TYPE_DEC);
+	if (decimal_is_neg(&mem->u.d)) {
+		int64_t i;
+		mem->type = MEM_TYPE_INT;
+		mem->flags = 0;
+		if (decimal_to_int64(&mem->u.d, &i) == NULL) {
+			mem->u.i = INT64_MIN;
+			return -1;
+		}
+		assert(i < 0);
+		mem->u.i = i;
+		/*
+		 * Decimal is floored when cast to int, which means that after
+		 * cast it becomes bigger if it was not integer.
+		 */
+		return decimal_is_int(&mem->u.d) ? 0 : -1;
+	}
+	uint64_t u;
+	mem->type = MEM_TYPE_UINT;
+	mem->flags = 0;
+	if (decimal_to_uint64(&mem->u.d, &u) == NULL) {
+		mem->u.u = UINT64_MAX;
+		return 1;
+	}
+	mem->u.u = u;
+	/*
+	 * Decimal is floored when cast to uint, which means that after cast it
+	 * becomes less if it was not integer.
+	 */
+	return decimal_is_int(&mem->u.d) ? 0 : 1;
+}
+
+static inline int
+dec_to_uint(struct Mem *mem)
+{
+	assert(mem->type == MEM_TYPE_DEC);
+	uint64_t u;
+	if (decimal_to_uint64(&mem->u.d, &u) == NULL)
+		return -1;
+	mem->u.u = u;
+	mem->type = MEM_TYPE_UINT;
+	mem->flags = 0;
+	return 0;
+}
+
+static inline int
+dec_to_uint_precise(struct Mem *mem)
+{
+	assert(mem->type == MEM_TYPE_DEC);
+	if (!decimal_is_int(&mem->u.d))
+		return -1;
+	return dec_to_uint(mem);
+}
+
+static inline int
+dec_to_uint_forced(struct Mem *mem)
+{
+	assert(mem->type == MEM_TYPE_DEC);
+	uint64_t u;
+	mem->type = MEM_TYPE_UINT;
+	mem->flags = 0;
+	if (decimal_to_uint64(&mem->u.d, &u) == NULL) {
+		if (decimal_is_neg(&mem->u.d)) {
+			mem->u.u = 0;
+			return -1;
+		}
+		mem->u.u = UINT64_MAX;
+		return 1;
+	}
+	mem->u.u = u;
+	/*
+	 * Decimal is floored when cast to uint, which means that after cast if
+	 * if was not integer it becomes less if it was positive, and move if it
```

Comment:
if -> it
Please add example to comments, it turns out to be hard to parse...

Reply:
Added.